### ISSUE TITLE
webgl

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP#234]
@abrahamjuliot - is this uptodate? https://gist.github.com/abrahamjuliot/7baf3be8c451d23f7a8693d7e28a35e2

I don't want to re-invent the wheel, so want to use your code (accredited) if that's OK - please advise

ATM, I only want to collect the parameters, extensions, vendor, etc - not any actual rendering

----

![webgl](https://user-images.githubusercontent.com/16656956/229646375-b6f982fb-4f3a-493c-bb5f-8d9df2132d8d.png)

so some questions

### Q1

https://gist.github.com/abrahamjuliot/7baf3be8c451d23f7a8693d7e28a35e2#file-webgl-js-L91-L94
```js
        let canvas = {}
        let context = {}
        try {
            canvas = document.createElement('canvas')
            context = canvas.getContext(type) || canvas.getContext('experimental-' + type)
```

so here, I wanted to use an already created canvas (I think it saves time)
```js
        let canvas = {}
	let context = {}
	try {
		canvas = dom.webglcanvas
		context = canvas.getContext(type) || canvas.getContext('experimental-' + type)
```

but I get errors for webgl2 (webgl is fine)

![errors](https://user-images.githubusercontent.com/16656956/229647153-581b9dad-c101-4956-a3b7-ef5b99d2af27.png)

I also don't see where you remove and cleanup the `document.createElement('canvas')` - remember, TZP allows section reruns so I don't want to end up with dozens of canvas elements

---

### Q2

what is newmethods meant to signify in webgl2 ?

---

### Q3

for data collection/entropy - I am not webgl savvy - but remember in the old TZP repo we discussed, and I mocked up, a webgl section that would return webgl, webgl2, experimental results - a bit like
- https://webglreport.com/ - webgl 1 + webgl2
- https://browserleaks.com/webgl - it has three toggles

how is this all reflected in the objects in the console? I'm a bit lost

---

### Q4

how/what are you doing with this data to only show one set of parameters on creepy - I can see that some are identical in webgl 1 vs 2 - and there is a diffs output. Are you merging these?

---

### Q5?

I guess I'll have to play to catch errors and return that - you seem to just console them and return undefined (I guess I could do that)

**Comment by Thorin-Oakenpants** [comment]
Q1 - answered: I just needed two canvas elements

```html
		<div>
			<canvas id="webglcanvas"></canvas>
			<canvas id="webgl2canvas"></canvas>
		</div>
```

is there some benefit to creating the canvas on the fly?

**Comment by abrahamjuliot** [comment]
> Q2

`newMethods` is just functions in `WebGLRenderingContext.prototype`. It's not that great for fingerprinting (compared to CSS and the Window). It just current browser features in the prototype.

**Comment by abrahamjuliot** [comment]
> Q3

I forgot to add the sections like "Vertex Shader" and "Rasterizer", but these can be determined based on the property names. We just need to create a map.

```js
parameterType = {
  'Vertex Shader': [
     'MAX_VERTEX_ATTRIBS',
     'MAX_VERTEX_TEXTURE_IMAGE_UNITS',
     'MAX_VERTEX_UNIFORM_VECTORS',
  ]
}

if (ParameterType['Vertex Shader'].includes(propertyName)) {
  // add to Vertex Shader section
}
```

**Comment by abrahamjuliot** [comment]
> Q4

I merge as a final simplification, but also check and flag mismatch behind the scenes. I recall an incident where a mismatch was valid, so I just treat it as questionable.

**Comment by Thorin-Oakenpants** [comment]
> I merge as a final simplification

hmmm, so I'm not sure how I want to present all this info - where is all the experimental stuff? I mean we can see webgl and webgl2. e.g. here's just checking for support. https://browserleaks.com/webgl has THREE toggles. Is this info in there (sorry for being a lazy bum but not into learning about webgl right now, lol) or it something we need to add

e.g.
```js
let types = ["webgl", "webgl2", "experimental-webgl"]
types.forEach(function(type){
	try {
		let canvas = window.document.createElement("canvas")
		try {
			var context = canvas.getContext(type)
			if (!context) {
				throw new Error()
			}
			console.log(type, "supported")
		}	catch(e) {
			console.log(type, "not supported")
		}
	} catch(e) {
		console.log("canvas failed")
	}
})
```

---

I think I can just take the objects and pull things out - I need to check various things as RFP protected, so they need to be separate.

> Since we are not appending the element to the page,

doh. nothing to remove. nice. BTW not creating the canvas is a tiny perf win :) woo!

**Comment by Thorin-Oakenpants** [comment]
https://github.com/arkenfox/TZP/issues/234#issuecomment-1495196774

can you add that to your gist so I can copy it, thanks

**Comment by abrahamjuliot** [comment]
Refactored: https://gist.github.com/abrahamjuliot/7baf3be8c451d23f7a8693d7e28a35e2

- added the option to check `experimental-webgl` separately
- added categories in final output
- added experimental concept `gpuHash`. The length may vary, so we can feed this to a SHA or mini hash instead of `join`
- dropped the supported functions since it is just browser feature detection
- added the major performance caveat detection
- added Direct3D detection

**Comment by Thorin-Oakenpants** [comment]
awesome, thx

**Comment by Thorin-Oakenpants** [comment]
the gpuBrand is redundant (but nice) - I can drop that

the gpuHash doesn't make much sense to me - sorting values could/would create collisions ? for example on it's own
- me: webgl - `NVIDIA:0:1:8:16:23:24:30:32:127:1024:4095:16384:32767`
- what do the numbers correspond to (we already have that data in parameters?)
   - e.g. `VERTEX_SHADER_BEST_FLOAT_PRECISION: Array(3) [ 23, 127, 127 ]`
- why a Set and only one of each number
   - see above I have two 127's

This doesn't make a good fingerprint for max entropy, but I get it, it's a nice wee string. Unless I'm missing something

**Comment by Thorin-Oakenpants** [comment]
`gpu` doesn't handle an empty string, returns `,`

**Comment by Thorin-Oakenpants** [comment]
I'm confused. Top is TB (which has RFP webgl mitigations - you get the same result in FF with RFP enabled), bottom is nightly with no RFP.

The bottom test looks fine, but _at first glance_ seem to have duplication, but the top test - why are we not returning a `debugRendererInfo` object with undefined items (and extensions that mimic this RFP behavior). Why is webglcontext.renderer different to gpuRenderer ?

https://bugzilla.mozilla.org/show_bug.cgi?id=1337157 - lemme look at this. I know it's supposed to return undefined or blanks and/or `Mozilla`

![i-am-a-bit-confused](https://user-images.githubusercontent.com/16656956/230488384-ad13f1d4-5f16-4abc-ba3a-b464e1ca8905.png)


```js
	const Categories = {
		'debugRendererInfo': [
			'UNMASKED_VENDOR_WEBGL',
			'UNMASKED_RENDERER_WEBGL',
		], // snip
	}

	parameters = { // snip
		UNMASKED_VENDOR_WEBGL: getUnmasked(context, 'UNMASKED_VENDOR_WEBGL'),
		UNMASKED_RENDERER_WEBGL: getUnmasked(context, 'UNMASKED_RENDERER_WEBGL')
	}
	parameters.DIRECT_3D = /Direct3D|D3D(\d+)/.test(parameters.UNMASKED_RENDERER_WEBGL)

	gpuVendor = parameters.UNMASKED_VENDOR_WEBGL
	gpuRenderer = parameters.UNMASKED_RENDERER_WEBGL
```

**Comment by Thorin-Oakenpants** [comment]
so `gpu` (which I split into gpuVendor and gpuRenderer)  is redundant then? right? It's a direct lookup of debuginfo? right?

**Comment by abrahamjuliot** [comment]
Yeah, gpu is non-essential. I forgot. GPU showing up under render is this issue. It can be ignored, but the way to around is to feature detect the version and then use renderer instead of debug info, but its only necessary to remove the console warning.

https://github.com/mdn/content/issues/12689

**Comment by abrahamjuliot** [comment]
> gpuHash 

This too is not needed. 

It can be useful to put known good fingerprints into a lookup table of (similar to known good audio sums). Clashing is likely and okay, as long random WebGL fingerprints find it hard to fit in. I'm guessing, based on limited data on CreepJS, the table size would require no more than 100 hashes. Everything unknown can be questioned until it is established trustworthy. This is more of a test concept I suppose.

**Comment by Thorin-Oakenpants** [comment]
tis all cool. The RFP notation should be pretty simple now I refreshed my memory as to what we were doing with it - there's more to it than just vendor/renderer. But easy to check the few places it shows.

> but its only necessary to remove the console warning

I log all errors as part of the error entropy :)

> gpu/gpuHash/gpuBrand

no worries. I get it was experimental - but this is gecko and all I care about is Tor Browser and RFP. If I check prototype lies (I'll need to check) then I can just return the whole thing as useless (might revisit later)

when we collect data from surveys (via a Tor Browser annual or bi-annual FP drive - 100% opt-in with one test per profile), we can just reject collecting tainted data based on prototype lies alone - i.e either it is all empty or it matches NoScripts signature

**Comment by abrahamjuliot** [comment]
There's also these anti-detection browsers that fake the GPUs at engine level. No prototype lies. For Firefox, I think they use what is called "Stealthfox Browser". The only way I'm aware of to detect them is by using a look-up table of known good hashes. WebGL  is too high entropy for a local lookup, but this lower entropy hash works. I have not used it on CreepJS, but I have the data visually from last 60 days of bad traffic.

**Comment by Thorin-Oakenpants** [comment]
gummy bear browsers :) lols ... I'm not too worried about it TBH, I need to focus on actual TB and FF + RFP users

**Comment by abrahamjuliot** [comment]
Something I discovered, many of them have a frozen max call stack fingerprint. Something with the way it's compiled, I guess.

**Comment by Thorin-Oakenpants** [comment]
I'm not entirely sure what you mean by that, tell me more :)

Yeah, math PoCs are great, like known pixel tests, joining chars vs individual chars in textmetrics width, domrect etc. Enumerating goodness is hard and no bulletproof - so I understand the mini simplified hash - like the simplified (less measurements) of https://arkenfox.github.io/TZP/tests/domrectspoofratio.html that covers multiple chrome results

I just had to drop enumerating goodness for audio in FF (TZP 2.0 is gecko only) - math library changes on android, and they will change again, and then RFP is also doing something with it (or it will apply for all) - https://bugzilla.mozilla.org/show_bug.cgi?id=1358149#c26

**Comment by Thorin-Oakenpants** [comment]
> you had me at "something"

do you mean recursion, stack depth, rather than some webgl thing?

**Comment by abrahamjuliot** [comment]
Yeah, recursion, stack depth. I have not tested for Firefox, but it's somewhat of an unbeatable fingerprint for custom Chrome builds that are slow to update.

**Comment by Thorin-Oakenpants** [comment]
OT: but I had an email exchange with a moz dev about this, and it's a good (fuzzy) FP for determining ion and jit etc

**Comment by Thorin-Oakenpants** [comment]
fyi: I want to get webgl and webgpu done soonish

@abrahamjuliot  ^ actually _your_ code has it .. the code you gave me didn't .. that sucks bro

Anyway, I was wondering how to display/structure it all. I see you combine it. I think I'll combine *vendor and *renderer into one, but then have webgl_1, webgl_2 and webgl_experimental all have a parameters and an extensions object.

soooo ... parameters - I see you list things quite differently - you can open TZP and look in the console (just imagine I put all the parameters into a single object) - then look at yours

```
FRAGMENT_SHADER.HIGH_FLOAT.precision: 23
FRAGMENT_SHADER.HIGH_FLOAT.rangeMax: 127
FRAGMENT_SHADER.HIGH_FLOAT.rangeMin: 127
FRAGMENT_SHADER.HIGH_INT.precision: 0
FRAGMENT_SHADER.HIGH_INT.rangeMax: 30
FRAGMENT_SHADER.HIGH_INT.rangeMin: 31
FRAGMENT_SHADER.LOW_FLOAT.precision: 23
FRAGMENT_SHADER.LOW_FLOAT.rangeMax: 127
FRAGMENT_SHADER.LOW_FLOAT.rangeMin: 127
FRAGMENT_SHADER.MEDIUM_FLOAT.precision: 23
FRAGMENT_SHADER.MEDIUM_FLOAT.rangeMax: 127
FRAGMENT_SHADER.MEDIUM_FLOAT.rangeMin: 127
```

whereas mine would be (webgl1) - that's it
```
FRAGMENT_SHADER_BEST_FLOAT_PRECISION: Array(3) [ 23, 127, 127 ]
```

what's that all about? unless I'm missing something (I'll have to do some sort of compare) but I see `MAX_CLIENT_WAIT_TIMEOUT_WEBGL: 1000000000` which isn't in mine.

also if you look at https://browserleaks.com/webgl - are we missing anything from there? I see

---

also whats your WebGPU code, thanks - spill it please. Also https://browserleaks.com/webgpu seems to have enumerated more properties but no values whilst you did `limits` only (with values) - can we ramp this sucker up a little

yours to save you looking
```
Limits
maxTextureDimension1D: 16384
maxTextureDimension2D: 16384
maxTextureDimension3D: 2048
maxTextureArrayLayers: 2048
maxBindGroups: 8
maxBindGroupsPlusVertexBuffers: 24
maxBindingsPerBindGroup: 65535
maxDynamicUniformBuffersPerPipelineLayout: 8
maxDynamicStorageBuffersPerPipelineLayout: 4
maxSampledTexturesPerShaderStage: 64
maxSamplersPerShaderStage: 64
maxStorageBuffersPerShaderStage: 64
maxStorageTexturesPerShaderStage: 64
maxUniformBuffersPerShaderStage: 64
maxUniformBufferBindingSize: 65536
maxStorageBufferBindingSize: 1073741824
minUniformBufferOffsetAlignment: 256
minStorageBufferOffsetAlignment: 32
maxVertexBuffers: 16
maxBufferSize: 1073741824
maxVertexAttributes: 32
maxVertexBufferArrayStride: 2048
maxInterStageShaderVariables: 16
maxColorAttachments: 8
maxColorAttachmentBytesPerSample: 32
maxComputeWorkgroupStorageSize: 16384
maxComputeInvocationsPerWorkgroup: 768
maxComputeWorkgroupSizeX: 1024
maxComputeWorkgroupSizeY: 1024
maxComputeWorkgroupSizeZ: 64
maxComputeWorkgroupsPerDimension: 65535 
```

**Comment by abrahamjuliot** [comment]
I see BrowserLeaks uses `context.getContextAttributes()` for WebGL. It's mainly the default attributes or what we configured for the context. Might find some extensions scripts that leak something there. Color space values look useful to compare with match media. `MAX_CLIENT_WAIT_TIMEOUT_WEBGL` It thinks it's a static constant and the same for all browsers, but I have not looked at it closely. `FRAGMENT_SHADER_BEST_FLOAT_PRECISION` as an array looks better.

```js
context = document.createElement('canvas').getContext('webgl')
context.getContextAttributes() // { alpha: true, ...
context.drawingBufferColorSpace // 'srgb'
context.unpackColorSpace // 'srgb'
```

Here's the gist of what I do for WebGPU, but I currently drop the features due to Blink being slightly unstable as Chrome 131+ (random features not showing). The order has significant entropy from what I observed. In a given release, different hardware has slightly different ordering and some subtle differences in support. I am not sure if Firefox exhibits the same.

```js
const getWebGpu = () => {
    if (!navigator.gpu) return Promise.resolve();
    
    return navigator.gpu.requestAdapter().then(adapter => {
        if (!adapter) return;
        
        const { limits, features, info } = adapter;
        
        // Latest implementations use info instead of requestAdapterInfo
        const adapterInfoPromise = info ? Promise.resolve(info) : adapter.requestAdapterInfo();
        
        return adapterInfoPromise.then(adapterInfo => {
            const limitsData = {};
            for (const prop in limits) {
                limitsData[prop] = limits[prop];
            }
            
            return {
                adapterInfo: {
                    vendor: adapterInfo.vendor,
                    architecture: adapterInfo.architecture,
                    description: adapterInfo.description,
                    device: adapterInfo.device,
                },
                limits: limitsData,
                features: [...features.values()],
            };
        });
    });
};

getWebGpu().then(console.log);
```

BrowserLeaks WebGPU page is capturing a lot more. Texture formats seem to be linked to the supported features. Here's a modification to get more of what they have. The complete listing of texture formats is from [here](https://gpuweb.github.io/gpuweb/#enumdef-gputextureformat). They perform additional layered checks on adapter information that does not appear useful. 

```js
const getWebGpu = () => {
    if (!navigator.gpu) return Promise.resolve();
    
    return navigator.gpu.requestAdapter().then(adapter => {
        if (!adapter) return;
        
        const { limits, features, info } = adapter;
        const adapterInfoPromise = info ? Promise.resolve(info) : adapter.requestAdapterInfo();
        
        return adapterInfoPromise.then(adapterInfo => {
            const limitsData = {};
            for (const prop in limits) {
                limitsData[prop] = limits[prop];
            }
                
            return adapter.requestDevice().then(device => {
                const supportedFormats = [];
                const testTextureFormat = format => {
                    try {
                        device.createTexture({
                            size: [1, 1],
                            format,
                            usage: GPUTextureUsage.RENDER_ATTACHMENT,
                        });
                        supportedFormats.push(format);
                    } catch {}
                };
                
                [
                    "rgba8unorm", "bgra8unorm", "rgba16float", "r8unorm", "r16float",
                    "rg8unorm", "rg16float", "rgba32float", "depth24plus", "depth32float"
                ].forEach(testTextureFormat);
                
                return {
                    adapterInfo: {
                        vendor: adapterInfo.vendor,
                        architecture: adapterInfo.architecture,
                        description: adapterInfo.description,
                        device: adapterInfo.device,
                    },
                    limits: limitsData,
                    features: [...features.values()],
                    preferredCanvasFormat: navigator.gpu.getPreferredCanvasFormat(),
                    supportedTextureFormats: supportedFormats,
                    wgslFeatures: navigator.gpu.wgslLanguageFeatures ? [...navigator.gpu.wgslLanguageFeatures.values()] : [],
                };
            });
        });
    });
};

getWebGpu().then(console.log);
```

**Comment by Thorin-Oakenpants** [comment]
nice, something for me to get into maybe next week

**Comment by Thorin-Oakenpants** [comment]
the second code snippet throws

![Image](https://github.com/user-attachments/assets/8a66bc43-b8d1-4c39-94bd-78c986d8ba1c)
